Dalamud/Networking/Pipes/Internal/LinkHandlerService.cs

```diff
@@ -0,0 +1,129 @@
+    /// </summary>
+    /// <param name="rpcHostService">The injected RPC host service.</param>
+    [ServiceManager.ServiceConstructor]
+    public LinkHandlerService(RpcHostService rpcHostService)
+    {
+        rpcHostService.AddMethod("handleLink", this.HandleLinkCall);
+    }
+
+    /// <inheritdoc/>
+    public void DisposeService()
+    {
+    }
+
+    /// <summary>
+    /// Register a handler for a namespace. All URIs with this namespace will be dispatched to the handler.
+    /// </summary>
+    /// <param name="ns">The namespace to use for this subscription.</param>
+    /// <param name="handler">The command handler.</param>
+    public void Register(string ns, Action<DalamudUri> handler)
+    {
+        if (string.IsNullOrWhiteSpace(ns))
+            throw new ArgumentNullException(nameof(ns));
+
+        var list = this.handlers.GetOrAdd(ns, _ => []);
+        lock (list)
+        {
+            list.Add(handler);
+        }
+
+        this.log.Verbose("Registered handler for {Namespace}", ns);
+    }
+
+    /// <summary>
+    /// Unregister a handler.
+    /// </summary>
+    /// <param name="ns">The namespace to use for this subscription.</param>
+    /// <param name="handler">The command handler.</param>
+    public void Unregister(string ns, Action<DalamudUri> handler)
+    {
+        if (string.IsNullOrWhiteSpace(ns))
+            return;
+
+        if (!this.handlers.TryGetValue(ns, out var list))
+            return;
+
+        lock (list)
+        {
+            list.RemoveAll(x => x == handler);
+        }
+
+        if (list.Count == 0)
+            this.handlers.TryRemove(ns, out _);
+
+        this.log.Verbose("Unregistered handler for {Namespace}", ns);
+    }
+
+    /// <summary>
+    /// Dispatch a URI to matching handlers.
+    /// </summary>
+    /// <param name="uri">The URI to parse and dispatch.</param>
+    public void Dispatch(DalamudUri uri)
+    {
+        this.log.Information("Received URI: {Uri}", uri.ToString());
+
+        var ns = uri.Namespace;
+        if (!this.handlers.TryGetValue(ns, out var list))
+            return;
+
+        Action<DalamudUri>[] snapshot;
+        lock (list)
+        {
+            snapshot = list.ToArray();
+        }
+
+        foreach (var h in snapshot)
+        {
+            try
+            {
+                h(uri);
+            }
+            catch (Exception e)
+            {
+                this.log.Warning(e, "Link handler threw for {UriPath}", uri.Path);
+            }
+        }
+    }
+
+    /// <summary>
+    /// The RPC-invokable link handler.
+    /// </summary>
+    /// <param name="uri">A plain-text URI to parse.</param>
+    public void HandleLinkCall(string uri)
+    {
+        if (string.IsNullOrWhiteSpace(uri))
+            return;
+
+        try
+        {
+            var du = DalamudUri.FromUri(uri);
+            this.Dispatch(du);
+        }
+        catch (Exception)
+        {
+            // swallow parse errors; clients shouldn't crash the host
```

Comment:
Log? Probably means that someone messed up and they need to figure out why

Reply:
Swapped to `#InvokeSafely()`.